### ISSUE TITLE
Linker produces colliding MVID for two different inputs

### ISSUE THREAD

**Issue by vitek-karas** [mono/linker#2203]
See https://github.com/dotnet/runtime/issues/45649 for more details.

This is inherently a Cecil issue, but tracking it here since we don't normally watch Cecil repo.

Adding to 6.0.0 for now - to consider this at least.

/cc @agocke

**Comment by vitek-karas** [comment]
https://github.com/dotnet/linker/pull/2384 brings the fix into `main`.
We should consider this for 6.0 servicing.

**Comment by filipnavara** [comment]
> We should consider this for 6.0 servicing.

Just a heads-up. We discussed an [unrelated issue on Discord](https://discord.com/channels/732297728826277939/732297965019988138/922411904851386408) and found out that assemblies in different Mono runtime packs now get different MVID despite identical IL code.

For reference, see these packages:
https://nuget.info/packages/Microsoft.NETCore.App.Runtime.Mono.osx-x64/6.0.1 https://nuget.info/packages/Microsoft.NETCore.App.Runtime.Mono.osx-arm64/6.0.1

Check it for System.Collections.dll, it is likely identical. Now try the same on 7.0 nightly builds and it's no longer identical.

**Comment by vitek-karas** [comment]
The new linker behavior should be:
* Identical input produces output with deterministic (meaning always the same) MVID
* Different input produces output with different MVID (this was not true before the fix as per above)

If anybody sees a different behavior then we need to look into that. So the question here is if the input to the linker was identical or not (note that identical IL is not enough, the above issue is that the dlls only differ in the architecture bit in the PE header, but otherwise have identical IL, those should still get different MVIDs).

The new linker also fixes the PDV checksum parts - so if for some reason the difference is only in the symbols, the dlls should not be identical (and should not get the same MVIDs).

**Comment by filipnavara** [comment]
I agree it should be checked carefully. Some of the assemblies are originally built as AnyCPU so I would expect them to be identical if everything is deterministic (including PDB paths). Note that Xamarin.Android apparently does some deduplication by MVID so the determinism across architectures should be tested somehow.

**Comment by jonathanpeppers** [comment]
I had to make a workaround for the Android workload in our .NET 7 PR:

https://github.com/xamarin/xamarin-android/commit/59a2630fe9f237107a93c9fe1bdcf0e4e6189c1c

Previously, many of the .NET assemblies output from the linker per architecture would have identical MVIDs. So for example `System.Console.dll` for android-arm64, android-x64, etc. would have identical MVIDs. The main assembly that differed in .NET 6 was `System.Private.CoreLib.dll`.

We use this info to put architecture specific assemblies in a subfolder in the final Android app, and so you end up with only a couple duplicated assemblies. What's happening now is every assembly is different, so I changed our logic (for now) to use file size instead of MVID. That seems to workaround this for now, but should we do something else instead?

**Comment by marek-safar** [comment]
> What's happening now is every assembly is different, so I changed our logic (for now) to use file size instead of MVID.

That sounds like a very bad idea. The same file size can hide so many differences especially with alignment.

**Comment by jonathanpeppers** [comment]
Yes, I definitely want to revert this. I hope before merging this PR?

**Comment by agocke** [comment]
@filipnavara Can we get those two binaries to check? We need to validate that the linker didn't regress this before we can backport to 6.0

**Comment by agocke** [comment]
@jonathanpeppers What's the status here? Are we clear to backport this to 6.0?

**Comment by agocke** [comment]
The change is already in for .NET 7, so it sounds like this would be a ship blocker for .NET 7. I agree that we should do the work in main first, then backport.
